cmdstanpy/model.py

```diff
@@ -330,30 +337,54 @@ def compile(
                     if self._compiler_options is not None:
                         cmd.extend(self._compiler_options.compose())
                     cmd.append(Path(exe_file).as_posix())
+
+                    sink = io.StringIO()
                     try:
-                        msg = do_command(cmd, cmdstan_path())
-                        if msg is not None and 'Warning or error:' in msg:
-                            msg = msg.split("Warning or error:", 1)[1].strip()
+                        do_command(cmd=cmd, cwd=cmdstan_path(), fd_out=sink)
+                    except RuntimeError as e:
+                        sink.write(f'\n{str(e)}\n')
+                        compilation_failed = True
+                    finally:
+                        console = sink.getvalue()
+
+                    if compilation_failed or 'Warning:' in console:
+                        lines = console.split('\n')
+                        warnings = [x for x in lines if x.startswith('Warning')]
+                        syntax_errors = [
+                            x for x in lines if x.startswith('Syntax error')
+                        ]
+                        semantic_errors = [
+                            x for x in lines if x.startswith('Semantic error')
+                        ]
+                        exceptions = [
+                            x
+                            for x in lines
+                            if x.startswith('Uncauth exception')
```

Comment:
```suggestion
                            if x.startswith('Uncaught exception')
```

We might also want to add messages that start with "Exception:" - this is what an internal ocaml exception would start with.